blas/impl/KokkosBlas2_ger_impl.hpp

```diff
@@ -0,0 +1,339 @@
+      , x_            (x)
+      , y_            (y)
+      , A_            (A)
+  {
+    static_assert(Kokkos::is_view<XViewType>::value, "XViewType must be a Kokkos::View.");
+    static_assert(Kokkos::is_view<YViewType>::value, "YViewType must be a Kokkos::View.");
+    static_assert(Kokkos::is_view<AViewType>::value, "AViewType must be a Kokkos::View.");
+
+    static_assert(static_cast<int>(XViewType::rank) == 1, "XViewType must have rank 1.");
+    static_assert(static_cast<int>(YViewType::rank) == 1, "YViewType must have rank 1.");
+    static_assert(static_cast<int>(AViewType::rank) == 2, "AViewType must have rank 2.");
+
+    static_assert(std::is_integral<IndexType>::value, "IndexType must be an integer.");
+  }
+
+  KOKKOS_INLINE_FUNCTION void operator()(const IndexType & i) const {
+    using KAT = Kokkos::ArithTraits<typename AViewType::non_const_value_type>;
+
+    if (alpha_ == KAT::zero()) {
+      // Nothing to do
+    }
+    else {
+      const IndexType    N      ( A_.extent(1) );
+      const A_value_type x_fixed( x_(i) );
+
+      if (justTranspose_) {
+        for (IndexType j = 0; j < N; ++j) {
+          A_(i,j) += A_value_type( alpha_ * x_fixed * y_(j) );
+        }
+      }
+      else {
+        for (IndexType j = 0; j < N; ++j) {
+          A_(i,j) += A_value_type( alpha_ * x_fixed * KAT::conj( y_(j) ) );
+        }
+      }
+    }
+  }
+
+private:
+  bool                           justTranspose_;
+  AlphaCoeffType                 alpha_;
+  typename XViewType::const_type x_;
+  typename YViewType::const_type y_;
+  AViewType                      A_;
+};
+
+// Single-level parallel version of GER.
+template <class XViewType, class YViewType, class AViewType,
+          class IndexType = typename AViewType::size_type>
+void singleLevelGer( const typename AViewType::execution_space  & space
+                   , const          char                          trans[]
+                   , const typename AViewType::const_value_type & alpha
+                   , const          XViewType                   & x
+                   , const          YViewType                   & y
+                   , const          AViewType                   & A
+                   ) {
+  KOKKOS_IMPL_DO_NOT_USE_PRINTF( "Entering IMPL singleLevelGer(), AViewType = %s\n", typeid(AViewType).name() );
+  static_assert(Kokkos::is_view<XViewType>::value, "XViewType must be a Kokkos::View.");
+  static_assert(Kokkos::is_view<YViewType>::value, "YViewType must be a Kokkos::View.");
+  static_assert(Kokkos::is_view<AViewType>::value, "AViewType must be a Kokkos::View.");
+
+  static_assert(static_cast<int>(XViewType::rank) == 1, "XViewType must have rank 1.");
+  static_assert(static_cast<int>(YViewType::rank) == 1, "YViewType must have rank 1.");
+  static_assert(static_cast<int>(AViewType::rank) == 2, "AViewType must have rank 2.");
+
+  static_assert(std::is_integral<IndexType>::value, "IndexType must be an integer");
+
```

Comment:
Can you add runtime checks of the consistency of the dimension of the vectors and the matrix?

Reply:
Now all these checks (including the dynamic ones) are happening only once at blas/src/KokkosBlas2_ger.hpp. Thanks.